README.md
# Sampension Design Tokens

Transforming tokens from Figma Tokens Studio to json, js and scss files.

Change your tokens in `tokens.json` (either directly or with the Figma Tokens plugin in Figma). The GitHub action will automatically generate tokens to the `dist/json` directory that can then be read by Style Dictionary, which will output tokens to the format you define in `style-dictionary.mjs`

## Generate tokens

1. `nvm use` to switch to the correct node version
2. `npm install` to install dependencies
3. `npm run build` to generate tokens

### Generate new theme tokens

To add a new theme go to `token.config.mjs` and list which sets to include in the new theme under `tokenSets`.

NB: The *shared* token set is mandatory.

```js
// example
"theme-dotcom": ["dotcom/semantic/typography", "dotcom/semantic/color"],
"theme-subbrand-a": ["subbrand-a/semantic/typography", "subbrand-a/semantic/color"],
```

### 🏃‍♀️ npm scripts

* `clean` - remove the `dist` directory
* `build-json:shared` - generate a json file for the *shared* tokens
* `build-js-scss:shared` - generate js and scss files for all tokens
* `build` - clean folders, then generate the tokens as json, js and scss

## 👷‍♀️ Workflow

* `tokens`-branch is used by the designers in the Figma Tokens plugin to pull and push tokens
* `main`-branch is used by the developers to generate the tokens npm package

When a designer is happy with their token changes they should make a pull request from `tokens` to `main` - if needed they should ask a developer for an introduction to pull requests.

## 🚀 Release

To release a new version, follow these steps:

1. Commit your changes to `main` branch - use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
2. A new version will automatically be released using `semantic-release` in a Github Action and published to npm.

## 👋 Contact

This project was set up by:

* Caroline Hansen: [caroline.hansen@manyone.com](mailto:caroline.hansen@manyone.com)
* Lau Rasmussen: [lau.rasmussen@manyone.com](mailto:lau.rasmussen@manyone.com)


CHANGELOG.md
## [1.2.34](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.33...v1.2.34) (2023-03-02)


### Bug Fixes

* **fonts:** make css font variables right ([2851e6d](https://github.com/manyone-cph/sampension-tokens/commit/2851e6dc2ab434a610a68ec838e8b3e60ff26737))

## [1.2.33](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.32...v1.2.33) (2023-02-08)


### Bug Fixes

* **font:** reference right default font ([b64d9ba](https://github.com/manyone-cph/sampension-tokens/commit/b64d9ba64414859ac318321cc00eaefcbd0c5b6f))

## [1.2.32](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.31...v1.2.32) (2023-02-08)


### Bug Fixes

* **section:** adjust section-sm size ([2796c42](https://github.com/manyone-cph/sampension-tokens/commit/2796c42fff0c3fc2c4c5b2945703205094150e94))

## [1.2.31](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.30...v1.2.31) (2023-02-08)


### Bug Fixes

* **section:** rename section to section-md + add section-sm ([6fd533f](https://github.com/manyone-cph/sampension-tokens/commit/6fd533f23cd6118094e97b464820969d69db5dd2))

## [1.2.30](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.29...v1.2.30) (2023-02-08)


### Bug Fixes

* **section:** reference dimension value ([ce0f9e5](https://github.com/manyone-cph/sampension-tokens/commit/ce0f9e5e4b98c0b71c359b656b8a89de586d06a5))

## [1.2.29](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.28...v1.2.29) (2023-02-08)

## [1.2.28](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.27...v1.2.28) (2023-02-07)


### Bug Fixes

* **release:** try even simpler ([2038c1f](https://github.com/manyone-cph/sampension-tokens/commit/2038c1f38c060fffbff8764fb3655876b076b93e))
* **release:** try to simplify stuff ([a5b3bdd](https://github.com/manyone-cph/sampension-tokens/commit/a5b3bddbc8a7253f5ffeb920c25121b47d12bcda))
* **release:** version number stuff ([51c8b0e](https://github.com/manyone-cph/sampension-tokens/commit/51c8b0e4750606973a1eb4c20619c7efa5bbb253))

## [1.2.27](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.26...v1.2.27) (2023-02-07)

## [1.2.26](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.25...v1.2.26) (2023-02-07)


### Bug Fixes

* **release:** module.exports in stead of module.export ([ecfd48d](https://github.com/manyone-cph/sampension-tokens/commit/ecfd48d27cfb404cc270d02cfb5b55ec53d1fe29))
* **release:** restructure files + make release config file ([9ab1db7](https://github.com/manyone-cph/sampension-tokens/commit/9ab1db750b735b3bc2e3ecdde8a0bbf0d48f85a5))
* **release:** try to make config file work ([32e262c](https://github.com/manyone-cph/sampension-tokens/commit/32e262cbd93b2eb07b84af0c97013ef824c16d15))

## [1.2.25](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.24...v1.2.25) (2023-02-07)


### Bug Fixes

* **gitignore:** ignore dist ([6864831](https://github.com/manyone-cph/sampension-tokens/commit/686483102cf2e303402e3a29a08e3b7bef2818d1))

## [1.2.24](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.23...v1.2.24) (2023-02-07)


### Bug Fixes

* **release:** copy package version to dist ([c249648](https://github.com/manyone-cph/sampension-tokens/commit/c249648a2e8e47d444241cf4b5a48b29c04dd861))
* **release:** dirname undefined ([165ffec](https://github.com/manyone-cph/sampension-tokens/commit/165ffece804e78c87e3458978bec50400a11ae95))

## [1.2.23](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.22...v1.2.23) (2023-02-07)


### Bug Fixes

* **release:** try to manually publish ([edaae2c](https://github.com/manyone-cph/sampension-tokens/commit/edaae2c0b7e0715bcf2fd0fa363daf88b8de60dc))

## [1.2.22](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.21...v1.2.22) (2023-02-07)


### Bug Fixes

* **release:** install missing dep ([0404963](https://github.com/manyone-cph/sampension-tokens/commit/04049638d2c3172049eb90ad6834ae8af004e30b))
* **release:** try again ([16bfefe](https://github.com/manyone-cph/sampension-tokens/commit/16bfefe7ad155ee115cd3a380f43845978d48709))
* **release:** try to hook in to publish command ([76ec4e8](https://github.com/manyone-cph/sampension-tokens/commit/76ec4e8db954e46a9d4e3f4ad882d9542527e697))

## [1.2.20](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.19...v1.2.20) (2023-02-07)


### Bug Fixes

* **release:** again ([a17024e](https://github.com/manyone-cph/sampension-tokens/commit/a17024e10049b15bdc936451b80a4cf4226e12a0))
* **release:** try again ([a4522be](https://github.com/manyone-cph/sampension-tokens/commit/a4522beb40394cb752367b86dc5704395109a329))
* **release:** try to bump package version and build nice package ([70268bd](https://github.com/manyone-cph/sampension-tokens/commit/70268bd7fbd3cf070009e765c29a6fb08e55a195))

## [1.2.19](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.18...v1.2.19) (2023-02-07)


### Bug Fixes

* **release:** build npm ([34bd3df](https://github.com/manyone-cph/sampension-tokens/commit/34bd3dffb0b78bd30289d918c191df20b1ed2ea1))

## [1.2.18](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.17...v1.2.18) (2023-02-07)


### Bug Fixes

* **release:** again ([8804d53](https://github.com/manyone-cph/sampension-tokens/commit/8804d535e2f1b050a7e9ade7ac53187ca4d5a13a))
* **release:** try again ([bfe6bca](https://github.com/manyone-cph/sampension-tokens/commit/bfe6bcafebf166a1f43d63d5da2ce4eb89897d1f))

## [1.2.17](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.16...v1.2.17) (2023-02-07)


### Bug Fixes

* **release:** try to make it work ([af34afb](https://github.com/manyone-cph/sampension-tokens/commit/af34afba7b08c3d285e2897fba4a5ce9fbc01ef5))

## [1.2.16](https://github.com/manyone-cph/sampension-tokens/compare/v1.2.15...v1.2.16) (2023-02-07)


### Bug Fixes

* **action:** bug with run and with ([96ef76d](https://github.com/manyone-cph/sampension-tokens/commit/96ef76d99f0ca1841b2808538a1108113135d4ba))

build-shared.js
// build.js
const StyleDictionary = require("style-dictionary");

// Convert to rem
const cleanRemSize = (val) => {
  // if the value is a string, remove the px and convert to rem
  if (typeof val == "string" && val.includes("px")) {
    val = Number(val.replace("px", ""));
  }

  if (typeof val == "number") {
    return val / 16 + "rem";
  } else {
    return val;
  }
};

// size/toREM -> Collection that needs converting to rem 
StyleDictionary.registerTransform({
  name: "size/toREM",
  type: "value",
  matcher: function (prop) {
    console.log(prop.type);
    return (
      prop.type === "fontSizes" ||
      prop.type === "lineHeights" || prop.type === "lineHeight" ||
      prop.type === "spacing" ||
      prop.type === "sizing" ||
      prop.type === "dimension" ||
      prop.type === "borderRadius" ||
      (prop.path.includes("typography") && prop.path.includes("fontSize")) ||
      (prop.path.includes("typography") && prop.path.includes("lineHeights"))
    );
  },
  transformer: function (prop) {
    return cleanRemSize(prop.value);
  },
});

// border/toShorthand -> Convert border object to border shorthand
StyleDictionary.registerTransform({
  name: "border/toShorthand",
  type: "value",
  matcher: function (prop) {
    return prop.type === "border";
  },
  transformer: function (prop) {
    return `${cleanRemSize(prop.value.width)} ${prop.value.style} ${prop.value.color}`;
  },
});

// fontWeights/toNumber -> Convert fontWeights to numbers
StyleDictionary.registerTransform({
  name: "fontWeights/toNumber",
  type: "value",
  matcher: function (prop) {
    return prop.type === "fontWeights";
  },
  transformer: function (prop) {
    // We could make this more robust by checking for all possible font weight values from figma
    switch (prop.value.toLowerCase()) {
      case "thin":
        return 100;
      case "extralight":
        return 200;
      case "light":
        return 300;
      case "regular":
        return 400;
      case "medium":
        return 500;
      case "bold":
        return 700;
      case "black":
        return 700;
      default:
        return 400;
    }
  },
});

// Convert fontFamily to css variable
StyleDictionary.registerTransform({
  name: "fontFamily/toCSSVariable",
  type: "value",
  matcher: function (prop) {
    return prop.type === "fontFamilies";
  },
  transformer: function (prop) {
    return `var(--ff-${prop.value.toLowerCase()})`;
  },
});



const filePaths = {
  shared: "tokens/shared.json"
};

const myStyleDictionary = StyleDictionary.extend({
  source: ["tokens/**/*.json"],
  platforms: {
    scss: {
      transformGroup: "scss",
      buildPath: "./dist/scss/",
      transforms: [
        "name/cti/kebab",
        "size/toREM",
        "border/toShorthand",
        "fontWeights/toNumber",
        "fontFamily/toCSSVariable",
      ],
      files: [
        {
          destination: "_shared.scss",
          format: "scss/variables",
          filter: (token) => {
            // only include: shared
            return token.filePath === filePaths.shared;
          },
        },
        {
          destination: "_shared_map.scss",
          format: "scss/map-deep",
          filter: (token) => {
            // only include: shared
            return token.filePath === filePaths.shared;
          },
        },
      ],
    },
    js: {
      transformGroup: "js",
      buildPath: "./dist/js/",
      files: [
        {
          destination: "_shared.js",
          format: "javascript/es6",
          filter: (token) => {
            // only include: shared
            return token.filePath === filePaths.shared;
          },
        },
      ],
    },
  },
});

myStyleDictionary.buildAllPlatforms();


release.config.js
module.exports = {
  branches: ["main"],
  plugins: [
    // 1. check if the commit should trigger a release
    [
      "@semantic-release/commit-analyzer",
      {
        preset: "angular",
        releaseRules: [
          {
            // release on all commits
            message: "*",
            release: "patch",
          },
        ],
      },
    ],
    // 2. generate release notes based on the commits
    "@semantic-release/release-notes-generator",
    // 3. update the changelog file with the release notes
    "@semantic-release/changelog",
    // 4. update the package.json with the new version
    [
      "@semantic-release/npm",
      {
        // we will publish manually in the next step
        npmPublish: false,
      },
    ],
    // 5. publish the package to npm from the dist folder
    [
      "@semantic-release/exec",
      {
        publishCmd: "cd ./dist && npm publish --access public",
      },
    ],
    // 6. commit the changes to the repository
    [
      "@semantic-release/git",
      {
        assets: ["package.json", "CHANGELOG.md"],
      },
    ],
    // 7. create a GitHub release
    "@semantic-release/github",
  ],
};


release.config.mjs
export default {
  branches: ["main"],
  plugins: [
    // 1. check if the commit should trigger a release
    [
      "@semantic-release/commit-analyzer",
      {
        preset: "angular",
        releaseRules: [
          {
            // release on all commits
            message: "*",
            release: "patch",
          },
        ],
      },
    ],
    // 2. generate release notes based on the commits
    "@semantic-release/release-notes-generator",
    // 3. update the changelog file with the release notes
    "@semantic-release/changelog",
    // 4. update the package.json with the new version
    [
      "@semantic-release/npm",
      {
        // we will publish manually in the next step
        npmPublish: false,
      },
    ],
    // 5. publish the package to npm
    [
      "@semantic-release/exec",
      {
        publishCmd: "cd ./dist && npm publish --access public",
      },
    ],
    // 6. commit the changes to the repository
    [
      "@semantic-release/git",
      {
        assets: ["package.json", "CHANGELOG.md"],
      },
    ],
    // 7. create a GitHub release
    "@semantic-release/github",
  ],
};


style-dictionary.mjs
// build.js
import StyleDictionary from "style-dictionary";

// Convert to rem
const cleanRemSize = (val) => {
  // if the value is a string, remove the px and convert to rem
  if (typeof val == "string" && val.includes("px")) {
    val = Number(val.replace("px", ""));
  }

  if (typeof val == "number") {
    return val / 16 + "rem";
  } else {
    return val;
  }
};

// size/toREM -> Collection that needs converting to rem
StyleDictionary.registerTransform({
  name: "size/toREM",
  type: "value",
  matcher: function (prop) {
    return (
      prop.type === "fontSizes" ||
      prop.type === "lineHeights" ||
      prop.type === "lineHeight" ||
      prop.type === "spacing" ||
      prop.type === "sizing" ||
      prop.type === "dimension" ||
      prop.type === "borderRadius" ||
      (prop.path.includes("typography") && prop.path.includes("fontSize")) ||
      (prop.path.includes("typography") && prop.path.includes("lineHeights"))
    );
  },
  transformer: function (prop) {
    return cleanRemSize(prop.value);
  },
});

// border/toShorthand -> Convert border object to border shorthand
StyleDictionary.registerTransform({
  name: "border/toShorthand",
  type: "value",
  matcher: function (prop) {
    return prop.type === "border";
  },
  transformer: function (prop) {
    return `${cleanRemSize(prop.value.width)} ${prop.value.style} ${prop.value.color}`;
  },
});

// fontWeights/toNumber -> Convert fontWeights to numbers
StyleDictionary.registerTransform({
  name: "fontWeights/toNumber",
  type: "value",
  matcher: function (prop) {
    return prop.type === "fontWeights";
  },
  transformer: function (prop) {
    // We could make this more robust by checking for all possible font weight values from figma
    switch (prop.value.toLowerCase()) {
      case "thin":
        return 100;
      case "extralight":
        return 200;
      case "light":
        return 300;
      case "regular":
        return 400;
      case "medium":
        return 500;
      case "bold":
        return 700;
      case "black":
        return 700;
      default:
        return 400;
    }
  },
});

// Convert fontFamily to css variable
StyleDictionary.registerTransform({
  name: "fontFamily/toCSSVariable",
  type: "value",
  matcher: function (prop) {
    return prop.type === "fontFamilies";
  },
  transformer: function (prop) {
    return `var(--ff-${prop.value.toLowerCase().replace(/ /g, "-")})`;
  },
});

const transforms = [
  "name/cti/kebab",
  "size/toREM",
  "border/toShorthand",
  "fontWeights/toNumber",
  "fontFamily/toCSSVariable",
];

export const generateStyles = async (jsonPath, setName) => {
  const styleDictionaryConfig = StyleDictionary.extend({
    source: [jsonPath],
    platforms: {
      js: {
        transformGroup: "js",
        buildPath: "dist/js/",
        files: [
          {
            destination: "shared.js",
            destination: `${setName}.js`,
            format: "javascript/es6",
          },
        ],
      },
      css: {
        transformGroup: "css",
        buildPath: "dist/css/",
        transforms,
        files: [
          {
            destination: `${setName}.css`,
            format: "css/variables",
          },
        ],
      },
      scss: {
        transformGroup: "scss",
        buildPath: "dist/scss/",
        transforms,
        files: [
          {
            destination: `${setName}.scss`,
            format: "scss/variables",
          },
          {
            destination: `${setName}-map.scss`,
            format: "scss/map-deep",
            mapName: `tokens-${setName}`,
          },
        ],
      },
      "scss-extend": {
        transformGroup: "css",
        buildPath: "dist/scss/",
        transforms,
        options: {
          selector: `%tokens-${setName}`,
        },
        files: [
          {
            destination: `${setName}-extend.scss`,
            format: "css/variables",
          },
        ],
      },
    },
  });

  await styleDictionaryConfig.buildAllPlatforms();
};


style-dictionary-dotcom.mjs
// build.js
import StyleDictionary from "style-dictionary";

// Convert to rem
const cleanRemSize = (val) => {
  // if the value is a string, remove the px and convert to rem
  if (typeof val == "string" && val.includes("px")) {
    val = Number(val.replace("px", ""));
  }

  if (typeof val == "number") {
    return val / 16 + "rem";
  } else {
    return val;
  }
};

// size/toREM -> Collection that needs converting to rem
StyleDictionary.registerTransform({
  name: "size/toREM",
  type: "value",
  matcher: function (prop) {
    return (
      prop.type === "fontSizes" ||
      prop.type === "lineHeights" ||
      prop.type === "lineHeight" ||
      prop.type === "spacing" ||
      prop.type === "sizing" ||
      prop.type === "dimension" ||
      prop.type === "borderRadius" ||
      (prop.path.includes("typography") && prop.path.includes("fontSize")) ||
      (prop.path.includes("typography") && prop.path.includes("lineHeights"))
    );
  },
  transformer: function (prop) {
    return cleanRemSize(prop.value);
  },
});

// border/toShorthand -> Convert border object to border shorthand
StyleDictionary.registerTransform({
  name: "border/toShorthand",
  type: "value",
  matcher: function (prop) {
    return prop.type === "border";
  },
  transformer: function (prop) {
    return `${cleanRemSize(prop.value.width)} ${prop.value.style} ${prop.value.color}`;
  },
});

// fontWeights/toNumber -> Convert fontWeights to numbers
StyleDictionary.registerTransform({
  name: "fontWeights/toNumber",
  type: "value",
  matcher: function (prop) {
    return prop.type === "fontWeights";
  },
  transformer: function (prop) {
    // We could make this more robust by checking for all possible font weight values from figma
    switch (prop.value.toLowerCase()) {
      case "thin":
        return 100;
      case "extralight":
        return 200;
      case "light":
        return 300;
      case "regular":
        return 400;
      case "medium":
        return 500;
      case "bold":
        return 700;
      case "black":
        return 700;
      default:
        return 400;
    }
  },
});

// Convert fontFamily to css variable
StyleDictionary.registerTransform({
  name: "fontFamily/toCSSVariable",
  type: "value",
  matcher: function (prop) {
    return prop.type === "fontFamilies";
  },
  transformer: function (prop) {
    return `var(--ff-${prop.value.toLowerCase()})`;
  },
});

const filePaths = {
  dotcom: "dist/json/dotcom.json",
};

const customStyleDictionary = StyleDictionary.extend({
  source: ["dist/json/*.json"],
  platforms: {
    extend: {
      transformGroup: "css",
      buildPath: "dist/css/",
      options: {
        selector: "%theme-dotcom",
      },
      transforms: [
        "name/cti/kebab",
        "size/toREM",
        "border/toShorthand",
        "fontWeights/toNumber",
        "fontFamily/toCSSVariable",
      ],
      files: [
        {
          destination: "dotcom.scss",
          format: "css/variables",
          filter: (token) => {
            // only include: dotcom
            return token.filePath === filePaths.dotcom;
          },
        },
      ],
    },
    variables: {
      transformGroup: "css",
      buildPath: "dist/css/",
      transforms: [
        "name/cti/kebab",
        "size/toREM",
        "border/toShorthand",
        "fontWeights/toNumber",
        "fontFamily/toCSSVariable",
      ],
      files: [
        {
          destination: "dotcom.css",
          format: "css/variables",
          filter: (token) => {
            // only include: dotcom
            return token.filePath === filePaths.dotcom;
          },
        },
      ],
    },
    scss: {
      transformGroup: "scss",
      buildPath: "dist/scss/",
      transforms: [
        "name/cti/kebab",
        "size/toREM",
        "border/toShorthand",
        "fontWeights/toNumber",
        "fontFamily/toCSSVariable",
      ],
      files: [
        {
          destination: "dotcom.scss",
          format: "scss/variables",
          filter: (token) => {
            // only include: dotcom
            return token.filePath === filePaths.dotcom;
          },
        },
        {
          destination: "dotcom-map.scss",
          format: "scss/map-deep",
          filter: (token) => {
            // only include: dotcom
            return token.filePath === filePaths.dotcom;
          },
        },
      ],
    },
    js: {
      transformGroup: "js",
      buildPath: "dist/js/",
      files: [
        {
          destination: "dotcom.js",
          format: "javascript/es6",
          filter: (token) => {
            // only include: dotcom
            return token.filePath === filePaths.dotcom;
          },
        },
      ],
    },
  },
});

export default customStyleDictionary;
